File: java/omnia/data/structure/mutable/Stack.java
package omnia.data.structure.mutable;

import omnia.contract.Countable;

import java.util.Iterator;
import java.util.Optional;

import static java.util.Objects.requireNonNull;

public interface Stack<E> extends Countable, Iterable<E> {

  void push(E item);

  Optional<E> pop();

  static <E> Stack<E> masking(java.util.Stack<E> javaStack) {
    return new Stack<>() {

      @Override
      public void push(E item) {
        javaStack.push(requireNonNull(item));
      }

      @Override
      public Optional<E> pop() {
        return javaStack.isEmpty() ? Optional.empty() : Optional.ofNullable(javaStack.pop());
      }

      @Override
      public int count() {
        return javaStack.size();
      }

      @Override
      public boolean isPopulated() {
        return !javaStack.isEmpty();
      }

      @Override
      public Iterator<E> iterator() {
        return new Iterator<>() {

          @Override
          public boolean hasNext() {
            return !javaStack.empty();
          }

          @Override
          public E next() {
            return javaStack.pop();
          }
        };
      }
    };
  }
}

File: src/omnia/data/structure/mutable/Stack.kt
package omnia.data.structure.mutable

import omnia.contract.Countable

interface Stack<E : Any> : Countable, Iterable<E> {

  fun push(item: E)

  fun pop(): E?

  fun peek(): E?

  companion object {

    fun <E : Any> masking(kotlinList: kotlin.collections.MutableList<E>): Stack<E> {
      return object : Stack<E> {
        override fun push(item: E) {
          kotlinList.add(item)
        }

        override fun pop(): E? {
          return kotlinList.removeLastOrNull()
        }

        override fun peek(): E? {
          return kotlinList.lastOrNull()
        }

        override fun count(): Int {
          return kotlinList.size
        }

        override val isPopulated: Boolean get() = kotlinList.isNotEmpty()

        override fun iterator(): Iterator<E> {
          return object : Iterator<E> {
            override fun hasNext(): Boolean {
              return kotlinList.isNotEmpty()
            }

            override fun next(): E {
              return kotlinList.removeLast()
            }
          }
        }
      }
    }
  }
}